xnmt/decoder.py

```diff
@@ -41,29 +45,67 @@ class MlpSoftmaxDecoder(RnnDecoder, Serializable):
 
   yaml_tag = u'!MlpSoftmaxDecoder'
 
-  def __init__(self, vocab_size, layers=1, input_dim=None, lstm_dim=None, mlp_hidden_dim=None, trg_embed_dim=None, dropout=None,
-               rnn_spec="lstm", residual_to_output=False):
+  def __init__(self, vocab_size, layers=1, input_dim=None, lstm_dim=None,
+               mlp_hidden_dim=None, trg_embed_dim=None, dropout=None,
+               rnn_spec="lstm", residual_to_output=False, input_feeding=False):
     super(MlpSoftmaxDecoder, self).__init__()
-    lstm_dim = lstm_dim or model_globals.get("default_layer_dim")
-    mlp_hidden_dim = mlp_hidden_dim or model_globals.get("default_layer_dim")
-    trg_embed_dim = trg_embed_dim or model_globals.get("default_layer_dim")
-    input_dim = input_dim or model_globals.get("default_layer_dim")
-    self.fwd_lstm = RnnDecoder.rnn_from_spec(rnn_spec, layers, trg_embed_dim, lstm_dim, model_globals.dynet_param_collection.param_col, residual_to_output)
-    self.mlp = mlp.MLP(input_dim + lstm_dim, mlp_hidden_dim, vocab_size, model_globals.dynet_param_collection.param_col)
+    # Define dim
+    lstm_dim       = model_globals.default_if_none(lstm_dim)
+    mlp_hidden_dim = model_globals.default_if_none(mlp_hidden_dim)
+    trg_embed_dim  = model_globals.default_if_none(trg_embed_dim)
+    input_dim      = model_globals.default_if_none(input_dim)
+    # Input feeding
+    self.input_feeding = input_feeding
+    self.lstm_dim = lstm_dim
+    lstm_input = trg_embed_dim
+    if input_feeding:
+      lstm_input += lstm_dim
+
+    # LSTM
+    self.fwd_lstm  = RnnDecoder.rnn_from_spec(spec       = rnn_spec,
+                                              num_layers = layers,
+                                              input_dim  = lstm_input,
+                                              hidden_dim = lstm_dim,
+                                              model = param_col(),
+                                              residual_to_output = residual_to_output)
+    # MLP
+    self.context_projector = linear.Linear(input_dim  = input_dim + lstm_dim,
+                                           output_dim = mlp_hidden_dim,
+                                           model = param_col())
+    self.vocab_projector = linear.Linear(input_dim = mlp_hidden_dim,
+                                         output_dim = vocab_size,
+                                         model = param_col())
+    # Dropout
     self.dropout = dropout or model_globals.get("dropout")
+    # Mutable state
     self.state = None
+    self.h_t = None
 
-  def initialize(self):
-    self.state = self.fwd_lstm.initial_state()
+  def initialize(self, h_f):
+    if not h_f:
+      raise ValueError("The encoder does not support get_final_state() yet. " +
+                       "Error in initializing decoder.")
+    state = self.fwd_lstm.initial_state()
+    state = state.set_s([h_f, dy.tanh(h_f)])
```

Comment:
Because h_f is the state of the LSTM in the encoder. I am passing this state as the first state of the LSTM in the decoder.